core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -322,6 +323,39 @@ public static long snapshotIdAsOfTime(Table table, long timestampMillis) {
     return snapshotId;
   }
 
+  /**
+   * Returns the ID of the most recent snapshot on the given branch
+   * as of the given time in milliseconds
+   *
+   * @param table a {@link Table}
+   * @param branch a {@link String}
+   * @param timestampMillis the timestamp in millis since the Unix epoch
+   * @return the snapshot ID
+   * @throws IllegalArgumentException when no snapshot is found in the table, on the given branch
+   * older than the timestamp
+   */
+  public static long snapshotIdAsOfTime(Table table, String branch, long timestampMillis) {
+    SnapshotRef ref = table.refs().get(branch);
+    Preconditions.checkArgument(ref != null, "Branch %s does not exist", branch);
+    Preconditions.checkArgument(ref.isBranch(), "Ref %s is a tag, not a branch", branch);
+    Long snapshotId = null;
+    long minimumTimeDifference = Long.MAX_VALUE;
+    for (Snapshot snapshot : ancestorsOf(ref.snapshotId(), table::snapshot)) {
+      if (snapshot.timestampMillis() <= timestampMillis) {
+        if (timestampMillis - snapshot.timestampMillis() <= minimumTimeDifference) {
+          minimumTimeDifference = timestampMillis - snapshot.timestampMillis();
+          snapshotId = snapshot.snapshotId();
+        }
+      }
+    }
+
+    Preconditions.checkArgument(snapshotId != null,
+        "Cannot find a snapshot older than %s on branch %s",
+        DateTimeUtil.formatTimestampMillis(timestampMillis), branch);
+
+    return snapshotId;
+  }
+
```

Comment:
Can we make reuse `snapshotIdAsOfTime` here? We will also need to make necessary changes for this function to accept a branch.

Reply:
I think we need a separate method which accepts the branch parameter (which is what is done here). This is because the existing method is a public utility, so we should not break compatibility there by requiring a branch parameter.

Reply:
We can let `snapshotIdAsOfTime(Table table, long timestampMillis)` call the new method right?

Reply:
@jackye1995 That's true, The existing snapshotIdAsOfTime can call the new method with the main branch.